File: test/extensions/language/optional_members/java/optional_members/OptionalArrayTest.java
package optional_members;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

import optional_members.optional_array.Data8;
import optional_members.optional_array.TestStruct;

public class OptionalArrayTest
{
    @Test
    public void data8()
    {
        final TestStruct test = new TestStruct();
        test.setHasData8(true);
        test.setData8(new Data8[4]);
        assertEquals(4, test.getData8().length);

        assertFalse(test.isData16Set());
        assertFalse(test.isData16Used());
        test.setData16(new short[4]);
        assertTrue(test.isData16Set());
        assertFalse(test.isData16Used());
    }

    @Test
    public void autoData8()
    {
        final TestStruct test = new TestStruct();
        assertFalse(test.isAutoData8Set());
        assertFalse(test.isAutoData8Used());
        test.setAutoData8(new Data8[4]);
        assertTrue(test.isAutoData8Set());
        assertTrue(test.isAutoData8Used());
        assertEquals(4, test.getAutoData8().length);
    }

    @Test
    public void data16()
    {
        final TestStruct test = new TestStruct();
        test.setHasData8(false);
        test.setData16(new short[4]);
        assertEquals(4, test.getData16().length);

        assertFalse(test.isData8Set());
        assertFalse(test.isData8Used());
        test.setData8(new Data8[4]);
        assertTrue(test.isData8Set());
        assertFalse(test.isData8Used());
    }

    @Test
    public void autoData16()
    {
        final TestStruct test = new TestStruct();
        assertFalse(test.isAutoData16Set());
        assertFalse(test.isAutoData16Used());
        test.setAutoData16(new short[4]);
        assertTrue(test.isAutoData16Set());
        assertTrue(test.isAutoData16Used());
        assertEquals(4, test.getAutoData16().length);
    }
}

File: test/extensions/language/optional_members/cpp/OptionalArrayTest.cpp
#include "gtest/gtest.h"
#include "optional_members/optional_array/TestStruct.h"
#include "zserio/RebindAlloc.h"

// just test setters and getters
namespace optional_members
{
namespace optional_array
{

using allocator_type = TestStruct::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

TEST(OptionalArrayTest, data8)
{
    TestStruct test;
    test.setHasData8(true);
    test.setData8(vector_type<Data8>(4));
    ASSERT_EQ(4, test.getData8().size());

    ASSERT_FALSE(test.isData16Set());
    ASSERT_FALSE(test.isData16Used());
    test.setData16(vector_type<int16_t>(4));
    ASSERT_TRUE(test.isData16Set());
    ASSERT_FALSE(test.isData16Used());
}

TEST(OptionalArrayTest, autoData8)
{
    TestStruct test;
    ASSERT_FALSE(test.isAutoData8Set());
    ASSERT_FALSE(test.isAutoData8Used());
    test.setAutoData8(vector_type<Data8>(4));
    ASSERT_TRUE(test.isAutoData8Set());
    ASSERT_TRUE(test.isAutoData8Used());
    ASSERT_EQ(4, test.getAutoData8().size());
}

TEST(OptionalArrayTest, data16)
{
    TestStruct test;
    test.setHasData8(false);
    test.setData16(vector_type<int16_t>(4));
    ASSERT_EQ(4, test.getData16().size());

    ASSERT_FALSE(test.isData8Set());
    ASSERT_FALSE(test.isData8Used());
    test.setData8(vector_type<Data8>(4));
    ASSERT_TRUE(test.isData8Set());
    ASSERT_FALSE(test.isData8Used());
}

TEST(OptionalArrayTest, autoData16)
{
    TestStruct test;
    ASSERT_FALSE(test.isAutoData16Set());
    ASSERT_FALSE(test.isAutoData16Used());
    test.setAutoData16(vector_type<int16_t>(4));
    ASSERT_TRUE(test.isAutoData16Set());
    ASSERT_TRUE(test.isAutoData16Used());
    ASSERT_EQ(4, test.getAutoData16().size());
}

} // namespace optional_array
} // namespace optional_members
